### PR TITLE
switch to catch2 v3

### LINKED ISSUE #6894: Does not build with current catch2
The internally shipped version of catch2 is stuck at 2.13.8, which upstream catch2 is at 3.4.0 and Linux distributions no longer ship that outdated version. Version 3.4.0 of catch2 no longer seems to ship `catch.hpp` (but `catch_all.hpp, see https://github.com/catchorg/Catch2/tree/devel/src/catch2), so the cmake check fails:

```
	cd obj-x86_64-linux-gnu && DEB_PYTHON_INSTALL_LAYOUT=deb cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DFETCHCONTENT_FULLY_DISCONNECTED=ON -DCMAKE_INSTALL_RUNSTATEDIR=/run -DCMAKE_SKIP_INSTALL_ALL_DEPENDENCY=ON "-GUnix Makefiles" -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu -DCMAKE_BUILD_TYPE=None -DCMAKE_SKIP_RPATH=ON -DRDK_INSTALL_INTREE=OFF -DRDK_INSTALL_STATIC_LIBS=OFF -DRDK_BUILD_THREADSAFE_SSS=ON -DRDK_BUILD_PYTHON_WRAPPERS=ON -DRDK_OPTIMIZE_POPCNT=OFF -DRDK_USE_URF=OFF -DRDK_INSTALL_COMIC_FONTS=OFF -DRDK_BUILD_CAIRO_SUPPORT=ON -DBoost_NO_BOOST_CMAKE=TRUE -DCMAKE_INSTALL_PREFIX=/usr -DCATCH_DIR=/usr/include/catch2 -DPYTHON_EXECUTABLE=/usr/bin/python3.11 ../ ..
-- The C compiler identification is GNU 13.2.0
-- The CXX compiler identification is GNU 13.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Catch not found in /usr/include/catch2
Downloading https://github.com/catchorg/Catch2/archive/v2.13.8.tar.gz...
CMake Error at External/catch/CMakeLists.txt:22 (file):
  file RENAME failed to rename

    /<<PKGBUILDDIR>>/External/catch/Catch2-2.13.8

  to

    /usr/include/catch2

  because: Invalid cross-device link
```

Please consider supporting a 3.x version of catch2 if you need to vendorize it, so that system-installed versions work as well.

### PR DESCRIPTION
**Pull request by greglandrum**
This just does a mechanical migration.
There are some things we can/should do to modernize our use of catch, but that can come in a later PR

some more info on possible modernization here: `https://github.com/catchorg/Catch2/blob/devel/docs/migrate-v2-to-v3.md#how-to-migrate-projects-from-v2-to-v3`

Fixes #6894

### CODE REVIEW

**Comment by greglandrum** [comment]
@mbanck could you please test to see if this resolves the problems you were seeing with the debian builds? 
I tried to setup the cmake file so that it will use a system version of catch2 if it's there, but can't test that myself since v3 of catch isn't easily available as a system package for me.

File: External/catch/CMakeLists.txt [hunk 1]
```diff
@@ -1,25 +1,11 @@
...
-  downloadAndCheckMD5("https://github.com/catchorg/Catch2/archive/v${RELEASE_NO}.tar.gz"
-        "${CMAKE_CURRENT_SOURCE_DIR}/v${RELEASE_NO}.tar.gz"
-        "3639fb4cb8020de0dcf1fd5addb39418")
-  execute_process(COMMAND ${CMAKE_COMMAND} -E tar zxf
-    ${CMAKE_CURRENT_SOURCE_DIR}/v${RELEASE_NO}.tar.gz
-    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
-  file(RENAME "Catch2-${RELEASE_NO}" "${CATCH_DIR}")
-  set(CATCH_INCLUDE_DIR "${CATCH_DIR}/single_include/catch2" PARENT_SCOPE)
-endif()
+FetchContent_Declare(
```
Comment (mbanck) [inline External/catch/CMakeLists.txt:3]: This does not work for downstream distributions, the general rule is that network-access during build is not allowed in order to be self-consistent and be reproducible.
So you can have the git download as a fallback, but rdkit should work fine with a system installed catch2

File: CMakeLists.txt [hunk 2]
```diff
@@ -150,8 +150,18 @@ set(RDKit_ExternalDir "${CMAKE_CURRENT_SOURCE_DIR}/External")
 set(RDKit_DataDir "${CMAKE_CURRENT_SOURCE_DIR}/Data")
 
 #include catch
-add_subdirectory(${RDKit_ExternalDir}/catch)
-message("CATCH: ${CATCH_INCLUDE_DIR}")
+find_package(catch2 3)
```
Comment (mbanck) [inline CMakeLists.txt:153]: This one needs to be `find_package(Catch2 3)`  (capitalized c) for me in order for CMake to pick things up, see `https://github.com/catchorg/Catch2/blob/devel/CMake/Catch2Config.cmake.in`

File: Code/GraphMol/FileParsers/file_parsers_catch.cpp [hunk 3]
```diff
@@ -5914,8 +5916,8 @@ M  END
 )CTAB"_ctab;
   REQUIRE(m);
 
-  // Currently, MaeMolSupplier does not recognize "either" double bonds, this just
```
Comment (mbanck) [inline Code/GraphMol/FileParsers/file_parsers_catch.cpp:5917]: Unrelated whitespace fixes (which made tha PR not apply cleanly for 2023_09_02 for me)

File: Code/GraphMol/MolHash/catch_tests.cpp [hunk 4]
```diff
@@ -733,19 +733,16 @@ TEST_CASE("GitHub Issue #6505") {
 
 TEST_CASE("Github Issue #6855 MakeScaffoldGeneric isotope removal") {
   SECTION("Extended Murcko") {
-    auto mol =
```
Comment (mbanck) [inline Code/GraphMol/MolHash/catch_tests.cpp:736]: Unrelated whitespace fixes (which made tha PR not apply cleanly for 2023_09_02 for me)

File: Code/GraphMol/FileParsers/file_parsers_catch.cpp (hunk 3 above)
Comment (greglandrum) [reply Code/GraphMol/FileParsers/file_parsers_catch.cpp:5917]: yeah, we do these automatically as files are touched. It's an unfortunate consequence of what is otherwise a good thing (automatic code reformattiing)

File: CMakeLists.txt (hunk 2 above)
Comment (greglandrum) [reply CMakeLists.txt]: > This one needs to be `find_package(Catch2 3)` (capitalized c) for me in order for CMake to pick things up, see `https://github.com/catchorg/Catch2/blob/devel/CMake/Catch2Config.cmake.in`

yep, fixed that typo

**Comment by mbanck** [comment]
Work fine for me now, thanks!

**Review by ricrogz — APPROVED** [review]
LGTM, just a small suggestion to remove a commented out line.

File: Code/cmake/Modules/RDKitUtils.cmake [hunk 5]
```diff
@@ -193,11 +193,10 @@ macro(rdkit_catch_test)
   CDR(RDKTEST_SOURCES ${RDKTEST_DEFAULT_ARGS})
   if(RDK_BUILD_CPP_TESTS)
     add_executable(${RDKTEST_NAME} ${RDKTEST_SOURCES})
-    target_include_directories(${RDKTEST_NAME} PRIVATE ${CATCH_INCLUDE_DIR})
-    target_link_libraries(${RDKTEST_NAME} rdkitCatch ${RDKTEST_LINK_LIBRARIES})
+    #target_include_directories(${RDKTEST_NAME} PRIVATE ${CATCH_INCLUDE_DIR})
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/cmake/Modules/RDKitUtils.cmake]: Done!
